File: src/commands/remove-tag-from-feature.ts
import { readFile, writeFile } from 'fs/promises';
import type { Command } from 'commander';
import { join } from 'path';
import chalk from 'chalk';
import * as Gherkin from '@cucumber/gherkin';
import * as Messages from '@cucumber/messages';

interface RemoveTagFromFeatureOptions {
  cwd?: string;
}

interface RemoveTagFromFeatureResult {
  success: boolean;
  valid: boolean;
  message?: string;
  error?: string;
}

export async function removeTagFromFeature(
  featureFilePath: string,
  tags: string[],
  options: RemoveTagFromFeatureOptions = {}
): Promise<RemoveTagFromFeatureResult> {
  const cwd = options.cwd || process.cwd();
  const filePath = join(cwd, featureFilePath);

  // Read feature file
  let content: string;
  try {
    content = await readFile(filePath, 'utf-8');
  } catch (error: any) {
    if (error.code === 'ENOENT') {
      return {
        success: false,
        valid: false,
        error: `File not found: ${featureFilePath}`,
      };
    }
    throw error;
  }

  // Parse Gherkin to get existing tags
  const uuidFn = Messages.IdGenerator.uuid();
  const builder = new Gherkin.AstBuilder(uuidFn);
  const matcher = new Gherkin.GherkinClassicTokenMatcher();
  const parser = new Gherkin.Parser(builder, matcher);

  let gherkinDocument;
  try {
    gherkinDocument = parser.parse(content);
  } catch (error: any) {
    return {
      success: false,
      valid: false,
      error: `Invalid Gherkin syntax: ${error.message}`,
    };
  }

  if (!gherkinDocument.feature) {
    return {
      success: false,
      valid: false,
      error: 'File does not contain a valid Feature',
    };
  }

  // Get existing feature-level tags
  const existingTags = gherkinDocument.feature.tags.map(t => t.name);

  // Check if tags exist
  for (const tag of tags) {
    if (!existingTags.includes(tag)) {
      return {
        success: false,
        valid: false,
        error: `Tag ${tag} not found on this feature`,
      };
    }
  }

  // Remove tags from the file
  const lines = content.split('\n');
  const tagsToRemove = new Set(tags);
  const filteredLines: string[] = [];

  for (const line of lines) {
    const trimmed = line.trim();
    // Skip lines that are tags we want to remove
    if (trimmed.startsWith('@') && tagsToRemove.has(trimmed)) {
      continue;
    }
    filteredLines.push(line);
  }

  const newContent = filteredLines.join('\n');

  // Validate the result is valid Gherkin
  let valid = true;
  try {
    const testBuilder = new Gherkin.AstBuilder(Messages.IdGenerator.uuid());
    const testMatcher = new Gherkin.GherkinClassicTokenMatcher();
    const testParser = new Gherkin.Parser(testBuilder, testMatcher);
    testParser.parse(newContent);
  } catch {
    valid = false;
  }

  // Write file
  await writeFile(filePath, newContent, 'utf-8');

  const tagList = tags.join(', ');
  return {
    success: true,
    valid,
    message: `Removed ${tagList} from ${featureFilePath}`,
  };
}

export async function removeTagFromFeatureCommand(
  featureFilePath: string,
  tags: string[]
): Promise<void> {
  try {
    const result = await removeTagFromFeature(featureFilePath, tags);

    if (!result.success) {
      console.error(chalk.red('Error:'), result.error);
      process.exit(1);
    }

    console.log(chalk.green(`✓ ${result.message}`));
    process.exit(0);
  } catch (error: any) {
    console.error(chalk.red('Error:'), error.message);
    process.exit(1);
  }
}

export function registerRemoveTagFromFeatureCommand(program: Command): void {
  program
    .command('remove-tag-from-feature')
    .description('Remove one or more tags from a feature file')
    .argument('<file>', 'Feature file path (e.g., spec/features/login.feature)')
    .argument('<tags...>', 'Tag(s) to remove (e.g., @deprecated @wip)')
    .action(async (file: string, tags: string[]) => {
      await removeTagFromFeatureCommand(file, tags);
    });
}

File: rust/fspec-core/src/commands/remove_tag_from_feature.rs
//! `remove-tag-from-feature` — Rust port of
//! `src/commands/remove-tag-from-feature.ts` (RPC-281).
//!
//! Removes one or more feature-level tags from a Gherkin feature file by
//! filtering out entire lines whose trimmed text exactly equals one of the
//! requested tag names. Mirrors the TypeScript implementation including:
//!
//!   - canonical error envelopes (`File not found`,
//!     `File does not contain a valid Feature`,
//!     `Tag <t> not found on this feature`);
//!   - whole-line equality removal (multi-tag lines like `@a @b` are NOT
//!     split — TS quirk preserved here intentionally so existing call sites
//!     get the same behaviour from either binary);
//!   - a success message of the form `Removed <tags> from <file>`.
//!
//! ## Two-front-doors
//! Both the LLM-facing dispatcher AND the standalone Rust binary's clap
//! subcommand call this single function (RPC-003 §7/§11). The CLI bridge at
//! `rust/fspec/src/remove_tag_from_feature.rs` is JSON marshalling only.

use std::collections::HashSet;
use std::path::Path;

use serde::Deserialize;
use serde_json::json;

use crate::error::FspecCoreError;
use crate::io::gherkin::parse_feature_lenient;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct RemoveTagFromFeatureArgs {
    file: String,
    tags: Vec<String>,
}

/// Dispatcher entry point.
pub async fn run(args_json: &str, project_root: &Path) -> Result<String, FspecCoreError> {
    let args: RemoveTagFromFeatureArgs =
        serde_json::from_str(args_json).map_err(|e| FspecCoreError::InvalidArgs {
            command: "remove-tag-from-feature",
            reason: format!("failed to parse args: {e}"),
        })?;

    let file_abs = project_root.join(&args.file);

    // ---- Read feature file ----
    let content = match std::fs::read_to_string(&file_abs) {
        Ok(s) => s,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
            return Err(FspecCoreError::InvalidArgs {
                command: "remove-tag-from-feature",
                reason: format!("File not found: {}", args.file),
            });
        }
        Err(source) => {
            return Err(FspecCoreError::Io {
                command: "remove-tag-from-feature",
                source,
            });
        }
    };

    // ---- Parse Gherkin (lenient parity with TS @cucumber/gherkin) ----
    let feature = match parse_feature_lenient(&content) {
        Ok(f) => f,
        Err(_) => {
            return Err(FspecCoreError::InvalidArgs {
                command: "remove-tag-from-feature",
                reason: "File does not contain a valid Feature".to_string(),
            });
        }
    };

    // ---- Existing feature-level tags ----
    // Gherkin 0.16's parser strips the leading `@` from each tag, so we
    // re-prepend it to compare against the canonical `@tag` form supplied
    // by callers and used everywhere else in fspec.
    let existing: Vec<String> = feature.tags.iter().map(|t| format!("@{t}")).collect();

    // ---- Existence gate ----
    for tag in &args.tags {
        if !existing.iter().any(|e| e == tag) {
            return Err(FspecCoreError::InvalidArgs {
                command: "remove-tag-from-feature",
                reason: format!("Tag {tag} not found on this feature"),
            });
        }
    }

    // ---- Whole-line equality filter (TS parity) ----
    let to_remove: HashSet<&str> = args.tags.iter().map(String::as_str).collect();
    let mut kept: Vec<&str> = Vec::new();
    for line in content.split('\n') {
        let trimmed = line.trim();
        if trimmed.starts_with('@') && to_remove.contains(trimmed) {
            continue;
        }
        kept.push(line);
    }
    let new_content = kept.join("\n");

    // ---- Validate result is still parseable Gherkin ----
    let valid = parse_feature_lenient(&new_content).is_ok();

    // ---- Write file ----
    std::fs::write(&file_abs, &new_content).map_err(|source| FspecCoreError::Io {
        command: "remove-tag-from-feature",
        source,
    })?;

    let tag_list = args.tags.join(", ");
    let response = json!({
        "success": true,
        "valid": valid,
        "message": format!("Removed {} from {}", tag_list, args.file),
    });

    serde_json::to_string(&response).map_err(|e| FspecCoreError::InvalidArgs {
        command: "remove-tag-from-feature",
        reason: format!("failed to serialise response: {e}"),
    })
}

#[cfg(test)]
mod tests {
    #![allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]
    use super::*;

    #[test]
    fn args_parse_camel_case() {
        let a: RemoveTagFromFeatureArgs =
            serde_json::from_str(r#"{"file":"spec/features/x.feature","tags":["@wip"]}"#).unwrap();
        assert_eq!(a.file, "spec/features/x.feature");
        assert_eq!(a.tags, vec!["@wip".to_string()]);
    }
}
